nomt/src/beatree/branch/node.rs

```diff
@@ -173,61 +169,63 @@ impl<'a> BranchNodeView<'a> {
 
 unsafe impl Send for BranchNode {}
 
-pub fn body_size(prefix_len: usize, separator_len: usize, n: usize) -> usize {
+pub fn body_size(prefix_len: usize, tot_separators_len: usize, n: usize) -> usize {
     // prefix plus separator lengths are measured in bits, which we round
-    // up to the next byte boundary and then follow by the node pointers.
-    (prefix_len + (separator_len * n) + 7) / 8 + (4 * n)
+    // up to the next byte boundary. They are preceded by cells and followed by node pointers
+    (n + 1) * 2 + (prefix_len + tot_separators_len + 7) / 8 + (n * 4)
 }
 
 pub struct BranchNodeBuilder {
     branch: BranchNode,
     index: usize,
     prefix_len: usize,
-    separator_len: usize,
+    cells: Vec<u8>,
```

Comment:
I don't think we should be allocating memory inside of the branch node builder, either here or with `separators`. 

You might want to look at the leaf builder / leaf gauge to see how the API might look for this.
With `BranchNodeBuilder::new(branch, n, prefix_len, sum_of_separator_lens)` you should be able to encode without doing any intermediate allocations.

Reply:
> Altering the gauge to produce sum_of_separator_lens_not_counting_prefix should be pretty straightforward. First you sum sum_of_separator_lens (not applying prefix compression). Then you subtract prefix_len * n from that value when calculating the body size.

I don't think this could work (given `sum_of_separator_lens` being gathered during ingestion in each gauge.ingest step)

The prefix_len could be bigger than some separator_len specified in `branch_gauge.ingest`, thus resulting in `sum_of_separator_lens - prefix_len * n` yielding fewer bytes than actually occupied.

I don't see how it could work without using something to store the separator lengths.
We could overestimate the total occupied bytes by simply summing `max(prefix_len, separator_len)` at each ingestion step.
However, the `prefix_len` may decrease over time, resulting in a size larger than the actual space that will be occupied in the branch node.

Reply:
> Altering the gauge to produce sum_of_separator_lens_not_counting_prefix should be pretty straightforward. First you sum sum_of_separator_lens (not applying prefix compression). Then you subtract prefix_len * n from that value when calculating the body size.

I don't think this could work (given `sum_of_separator_lens` being gathered during ingestion in each gauge.ingest step)

The prefix_len could be bigger than some separator_len specified in `branch_gauge.ingest`, thus resulting in `sum_of_separator_lens - prefix_len * n` yielding fewer bytes than actually occupied.

I don't see how it could work without using something to store the separator lengths.
We could overestimate the total occupied bytes by simply summing `max(prefix_len, separator_len)` at each ingestion step.
However, the `prefix_len` may decrease over time, resulting in a size larger than the actual space that will be occupied in the branch node.

I tested this by making a small change to `tot_separators_len`

```rust
    fn tot_separators_len(&self) -> usize {
        let r1 = self
            .separators_len
            .iter()
            .map(|s| s.saturating_sub(self.prefix_len))
            .sum();
        let sum_of_separator_lens_not_counting_prefix = self.separators_len.iter().sum::<usize>();
        let r2 =  sum_of_separator_lens_not_counting_prefix - (self.prefix_len * self.n);
        if self.n != 1 {
            assert_eq!(r1, r2); // this fails immediately 
        }
        r1
    }
```

In the first iteration of `cargo run --release -- run -w transfer -c 20 -b nomt --time-limit 60s -s 10000 -r --buckets 500000` with `n=2` we have `&self.separators_len = [1, 17,]` and `self.prefix_len = 13`, which results in a negative outcome in this case

Reply:
> Altering the gauge to produce sum_of_separator_lens_not_counting_prefix should be pretty straightforward. First you sum sum_of_separator_lens (not applying prefix compression). Then you subtract prefix_len * n from that value when calculating the body size.

I don't think this could work (given `sum_of_separator_lens` being gathered during ingestion in each gauge.ingest step)

The prefix_len could be bigger than some separator_len specified in `branch_gauge.ingest`, thus resulting in `sum_of_separator_lens - prefix_len * n` yielding fewer bytes than actually occupied.

I don't see how it could work without using something to store the separator lengths.

I tested this by making a small change to `tot_separators_len`

```rust
    fn tot_separators_len(&self) -> usize {
        let r1 = self
            .separators_len
            .iter()
            .map(|s| s.saturating_sub(self.prefix_len))
            .sum();
        let sum_of_separator_lens_not_counting_prefix = self.separators_len.iter().sum::<usize>();
        let r2 =  sum_of_separator_lens_not_counting_prefix - (self.prefix_len * self.n);
        if self.n != 1 {
            assert_eq!(r1, r2); // this fails immediately 
        }
        r1
    }
```

In the first iteration of `cargo run --release -- run -w transfer -c 20 -b nomt --time-limit 60s -s 10000 -r --buckets 500000` with `n=2` we have `&self.separators_len = [1, 17,]` and `self.prefix_len = 13`, which results in a negative outcome in this case

Reply:
I see. I hadn't considered the case where prefix_len was greater than the separator len. I suppose that if we have two keys:

`10000...` (total separator length 1)
`100001...` (total separator length 6)

the prefix length will be 5.

> I don't see how it could work without using something to store the separator lengths.

I can understand how you arrived at this conclusion based off of this edge case. However, I think we should be able to overcome this behavior by setting `separator_len` to `max(prefix_len, separator_len)`. It is always safe to logically zero-extend separators. The encoding of the branch with the example above would be

`[prefix: 10000][separator 1: len 0|][separator 2: len 1|1]` 

This will still require something to store separator lengths, but hopefully we can just reuse the `ops` Vec in the branch updater for that purpose.

I don't view that as a blocker for this PR and I am OK with merging it as-is and addressing later.

Reply:
The other option is to find a way to reuse the allocation. I don't think that copying is avoidable here. We just shouldn't be allocating memory on every branch we build!

Reply:
>It's worth noting that only the very first key can have separator_len <= prefix_len.

Yes, that's absolutely true. It is not only worth noting, but it is the key point that enables us to address the problem effectively and cleanly without any allocation

> When you push a new item, the prefix length must either reduce or stay the same

The second push could cause the prefix_len to increase, from the third push onwards, it can only reduce or remain the same.
However, this is not a problem because we simply store the length of the first separator.
In fact, in my previous code, the issue was solely with the *first* separator length.

The following code works perfectly fine:

```rust
    fn total_separators_len(&self) -> usize {
        let r1 = self
            .separator_lens
            .iter()
            .map(|s| s.saturating_sub(self.prefix_len))
            .sum();
        let r2 = self.first_separator.unwrap().1 // first separator_len
            .saturating_sub(self.prefix_len)
            + self.sum_separator_lens - (self.n - 1) * self.prefix_len;
        if r1 != r2 {
            panic!();
        }
        r2
    }
```

where `sum_separator_lens` is increased each ingest step (beside the first one) in the following manner
`self.sum_separator_lens += separator_len;`

>When you push a new item, the prefix length must either reduce or stay the same. You can increase prefix_compressed_separator_lengths by prefix_bits_lost * prefix_compressed_items.

I don't understand what you mean by 'prefix_bits_lost' here.

Anyway, I just pushed updated code without allocations in the gauge :fire:  (the new code also contains a correction of the method body_size_after)

Reply:
> I don't understand what you mean by 'prefix_bits_lost' here.

The change in the prefix len from pushing one key to the next